comp-bot/src/main/java/frc/robot/lights/LightsState.java

```diff
@@ -1,52 +1,82 @@
+  INTAKING_CORAL(Color.kWhite, 0.25),
+  INTAKING_ALGAE(Color.kTeal, 0.25),
 
-  IDLE_EMPTY(Color.kBlack, BlinkPattern.SOLID),
-  HOLDING_CORAL(Color.kWhite, BlinkPattern.SOLID),
-  HOLDING_ALGAE(Color.kTeal, BlinkPattern.SOLID),
+  IDLE_EMPTY(Color.kBlack, 0.0),
+  HOLDING_CORAL(Color.kWhite, 0.0),
+  HOLDING_ALGAE(Color.kTeal, 0.0),
 
-  LOLLIPOP_SEES_ALGAE(Color.kTeal, BlinkPattern.SOLID),
-  LOLLIPOP_NO_ALGAE(Color.kBlack, BlinkPattern.SOLID),
+  LOLLIPOP_SEES_ALGAE(Color.kTeal, 0.0),
+  LOLLIPOP_NO_ALGAE(Color.kBlack, 0.0),
 
-  CORAL_HANDOFF(Color.kWhite, BlinkPattern.BLINK_FAST),
+  CORAL_HANDOFF(Color.kWhite, 0.08),
 
-  CLIMB_LINEUP(Color.kYellow, BlinkPattern.SOLID),
-  CLIMB_HANG(Color.kGreen, BlinkPattern.SOLID),
-  CLIMB_STOP(Color.kGreen, BlinkPattern.BLINK_SLOW),
+  CLIMB_LINEUP(Color.kYellow, 0.0),
+  CLIMB_HANG(Color.kGreen, 0.0),
+  CLIMB_STOP(Color.kGreen, 0.25),
 
-  SCORE_NO_ALIGN_NO_TAGS(Color.kYellow, BlinkPattern.BLINK_SLOW),
-  SCORE_NO_ALIGN_TAGS(Color.kGreen, BlinkPattern.BLINK_SLOW),
-  SCORE_ALIGN_NO_TAGS(Color.kYellow, BlinkPattern.BLINK_FAST),
-  SCORE_ALIGN_TAGS(Color.kGreen, BlinkPattern.BLINK_FAST),
+  SCORE_NO_ALIGN_NO_TAGS(Color.kYellow, 0.25),
+  SCORE_NO_ALIGN_TAGS(Color.kGreen, 0.25),
+  SCORE_ALIGN_NO_TAGS(Color.kYellow, 0.08),
+  SCORE_ALIGN_TAGS(Color.kGreen, 0.08),
 
-  SCORING_ALGAE(Color.kTeal, BlinkPattern.BLINK_FAST),
-  SCORING_CORAL(Color.kWhite, BlinkPattern.BLINK_FAST),
+  SCORING_ALGAE(Color.kTeal, 0.08),
+  SCORING_CORAL(Color.kWhite, 0.08),
 
-  OTHER(Color.kPurple, BlinkPattern.BLINK_SLOW),
+  OTHER(Color.kPurple, 0.25),
 
   /**
    * @deprecated Replace placeholder lights with actual light patterns.
    */
   @Deprecated
-  PLACEHOLDER(Color.kBlack, BlinkPattern.SOLID);
+  PLACEHOLDER(Color.kBlack, 0.0);
+
+  public final Optional<StrobeAnimation> stateBlinkRequest;
+  public final Optional<SolidColor> stateColorRequest;
 
-  public final BlinkPattern pattern;
   public final Color color;
+  public final double duration;
 
-  LightsState(Color color, BlinkPattern pattern) {
-    this.pattern = pattern;
+  LightsState(Color color, double duration) {
     this.color = color;
+    this.duration = duration;
+    if (duration == 0.0) {
+      this.stateBlinkRequest = Optional.empty();
+      this.stateColorRequest =
+          Optional.of(
+              new SolidColor(0, 399).withColor(new RGBWColor(color)).withUpdateFreqHz(50.0));
+    } else {
+      this.stateBlinkRequest =
+          Optional.of(
+              new StrobeAnimation(0, 399)
+                  .withColor(new RGBWColor(color))
+                  .withFrameRate(1 / duration)
+                  .withUpdateFreqHz(50.0));
+      this.stateColorRequest = Optional.empty();
+    }
+  }
+
+  public boolean blinks() {
+    return stateBlinkRequest.isPresent();
+  }
+
+  public ControlRequest getControlRequest() {
+    return blinks() ? stateBlinkRequest.orElseThrow() : stateColorRequest.orElseThrow();
   }
```

Comment:
when you broaden the return type to a generic `ControlRequest`, it prevents the compiler from being able to optimize the call:

<img width="1492" height="1588" alt="Image" src="https://github.com/user-attachments/assets/1286b934-d784-42c0-9777-9b24dc2859c0" />

it will run each of those if statements one at a time until it can figure out what `ControlRequest` subclass to use. can we instead move this logic inside `LightsSubsystem`? it can do
```
if (blink state is present) {
  candle.setControl(blink state);
} else {
  candle.setControl(solid color state);
}
```

this will enable the compiler to optimize this, since it knows it'll either be using a blink request OR a solid color request. very small change but this kind of stuff can add up!